### ISSUE TITLE
optimizer weights not saved while keras.model.fit with colab tpu

### ISSUE THREAD

**Issue by SchenbergZY** [tensorflow/tpu#341]
After model.fit for 1 epoch I saved my model by tf.keras.model.save. 
But when I open the h5 file by import hdf5 I just saw 'model_weights' without 'optimizer_weights' at the same time.
Maybe someone can explain (or maybe fix?) this.. Thanks I really need this.

PS: I want to fine-tune whole BERT large via colab TPU but the maximum running time is 12 
      hours..
      For each epoch I need approx. 15000 seconds and 12 hours is not enough for 5 epochs. So, if 
      there are any advices, I will be really appreciated.

**Comment by saberkun** [comment]
BERT code was written using Estimator + basic TF. I guess keras model may not capture the TF optimizer correctly. We have tfhub example for finetuning with TPUEstimator. Could you try that?https://github.com/google-research/bert/blob/master/run_classifier_with_tfhub.py

**Comment by SchenbergZY** [comment]
> BERT code was written using Estimator + basic TF. I guess keras model may not capture the TF optimizer correctly. We have tfhub example for finetuning with TPUEstimator. Could you try that?https://github.com/google-research/bert/blob/master/run_classifier_with_tfhub.py

Well this is true but I need to change a lot like loss functions and output layers..
Also my data is 2million*200len, which causes OOM on Colab, so I need to split data into folds and train it folds by folds. I can't find a way to do this with estimator.train while model.fit can easily handle this by just fit multiple times. That's the reason I use keras. OR if you have any suggestion.

**Comment by saberkun** [comment]
Do you use tf.data to build input pipeline? Setting a smaller buffer and less prefetch will reduce memory usage. For keras, if you fit with numpy, it will definitely cause problem. For both estimator and keras, you should be able to use dataset (tf.data) to process. It is just recommended to work with TPUEstimator for TPU given the current TF version.
